transport/ProbeServer.cpp

```diff
@@ -0,0 +1,224 @@
+#include "transport/ProbeServer.h"
+#include "logger/Logger.h"
+#include "transport/ice/Stun.h"
+#include "utils/Time.h"
+#include <unistd.h>
+
+namespace transport
+{
+ProbeServer::ProbeServer(ice::IceConfig& iceConfig) : _iceConfig(iceConfig)
+{
+    // TODO copied from IceSession.cpp; refactor
+    char ufrag[14 + 1];
+    char pwd[24 + 1]; // length selected to make attribute *4 length
+
+    generateCredentialString(ufrag, sizeof(ufrag) - 1);
+    generateCredentialString(pwd, sizeof(pwd) - 1);
+
+    _probeCredentials = std::make_pair<std::string, std::string>(ufrag, pwd);
+}
+
+// Endpoint::IEvents
+void ProbeServer::onRtpReceived(Endpoint& endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet)
+{
+}
+
+void ProbeServer::onDtlsReceived(Endpoint& endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet){};
+
+void ProbeServer::onRtcpReceived(Endpoint& endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet)
+{
+}
+
+void ProbeServer::onIceReceived(Endpoint& endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet)
+{
+    replyStunOk(endpoint, source, std::move(packet));
+}
+
+void ProbeServer::onRegistered(Endpoint& endpoint)
+{
+    const std::lock_guard<std::mutex> locker(_endpointsLock);
+
+    if (endpoint.getTransportType() != ice::TransportType::UDP)
+    {
+        return;
+    }
+
+    auto port = endpoint.getLocalPort().getPort();
+    auto it = _udpPorts.find(port);
+    if (it != _udpPorts.end())
+    {
+        return;
+    }
+
+    if (!_iceConfig.publicIpv4.empty() && endpoint.getLocalPort().getFamily() == AF_INET)
+    {
+        auto address = _iceConfig.publicIpv4;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::UDP, address});
+
+        _udpPorts.insert(port);
+
+        logger::debug("Registered UDP endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+
+    if (!_iceConfig.publicIpv6.empty() && endpoint.getLocalPort().getFamily() == AF_INET6)
```

Comment:
The publicIpv6 config may be empty. TransportImpl still adds the local IP as a candidate, which in case of IPv6 is the same as the public. 
For ipv4 this scheme may work as the private IPv4 is not much point in listing as candidate when client cannot reach it. However, if SMB would be placed on LAN at customer, the local ipv4 is of interest. Perhaps we should not sut that door. SMB does expose the private IP to the clients joining, so this should do the same out of symetry.

Reply:
So we should suggest both ipv4 and ipv6 local IPs as probe candidates as well?

Reply:
So we should suggest both ipv4 and ipv6 local IPs as probe candidates as well?

Reply:
Updated to add candidates with local IPs (HOST type).